File: server/mongodb.ts
const { MongoClient, ObjectId } = require("mongodb");

function recreateID( array ) {
  for (let i = 0; i < array.length; i++)
    if (array[i] != undefined)
      if (array[i]._id != undefined)
        array[i]._id = new ObjectId(array[i]._id);//delete array[i]._id;
  return array;
}

class Database {
  nodesC; // Nodes collection
  connectionsC; // Connection collection
  varsC; // Variables collection
  
  async init( db ) {
    this.nodesC = await db.collection("nodes");
    this.connectionsC = await db.collection("connections");
    this.varsC = await db.collection("variables");
  }

  //////////// Nodes

  async getNode( uri ) {
    var node = await this.nodesC.findOne({_id: new ObjectId(uri)});
    return node;
  }

  async updateNode( uri, newData ) {
    var result = await this.nodesC.updateOne({_id: new ObjectId(uri)}, { $set: newData });
    return result;
  }

  async delNode( uri ) {
    let uriStr = `[${[...uri.id]}]`;

    return Promise.all([
      this.nodesC.deleteOne({_id: new ObjectId(uri)}),
      this.connectionsC.deleteMany({ id1: uriStr }),
      this.connectionsC.deleteMany({ id2: uriStr }),
    ]);
  } /* delNode */
  
  async addNode( data ) {
    var insertedURI = (await this.nodesC.insertOne(data)).insertedId;
    return insertedURI.id;
  } /* addNode */

  async getNodeConnections( uri ) {
    let uriStr = "[" + new Uint8Array(uri).toString() + "]";

    let cs = (await (await this.connectionsC.find({ id1: uriStr })).toArray()).concat((await (await this.connectionsC.find({ id2: uriStr })).toArray()));
    
    return cs;
  }

  async getAllConnections() {
    let cs = await (await this.connectionsC.find({})).toArray();
    
    return cs;
  }

  async getNeighbours( uri ) {
    let uriStr = "[" + new Uint8Array(uri).toString() + "]";

    let right = await (await this.connectionsC.find({ id1: uriStr })).toArray();
    let left = await (await this.connectionsC.find({ id2: uriStr })).toArray();

    let outN = [];
    for (let i = 0; i < right.length; i++)
      outN[outN.length] = right[i].id2;
    for (let i = 0; i < left.length; i++)
      outN[outN.length] = left[i].id1;
    return outN;
  }

  //////////////////// Connections

  async addConnection( uri1, uri2 ) {
    if (uri1 == uri2) // Check for connecting node with itself
      return; 

    let uri1Str = "[" + new Uint8Array(uri1).toString() + "]";
    let uri2Str = "[" + new Uint8Array(uri2).toString() + "]";

    let n1 = this.getNode(uri1);
    let n2 = this.getNode(uri2);
  
    if (n1 == null || n2 == null) // Check for nodes validity
      return;

    var c1 = await this.connectionsC.findOne({ id1: uri1Str, id2: uri2Str }),
        c2 = await this.connectionsC.findOne({ id1: uri2Str, id2: uri1Str });

    if (c1 != null || c1 != null) // Check for connections existing
      return false;

    var insertedURI = (await this.connectionsC.insertOne({ id1: uri1Str, id2: uri2Str })).insertedId;
    console.log("Inserted URI: ");
    console.log(insertedURI);
    return true;
  }

  async delConnection( uri1, uri2 ) {
    let uri1Str = "[" + new Uint8Array(uri1).toString() + "]";
    let uri2Str = "[" + new Uint8Array(uri2).toString() + "]";

    // Delete all connections
    this.connectionsC.deleteOne({ id1: uri1Str, id2: uri2Str });
    this.connectionsC.deleteOne({ id1: uri2Str, id2: uri1Str });
    return true;
  }

  ////////////// Indexes 

  async getDefNodeURI() {
    var node = await this.varsC.findOne({var_name: "DefNodeURI" });
    if (node != null)
      return node.uri.buffer;
    return null;
  }
  
  async setDefNodeURI( newURI ) {
    if ((await this.getNode(newURI)) == null)
      return false;

    var node = await this.varsC.findOne({var_name: "DefNodeURI" });
    if (node == null)
    {
      await this.varsC.insertOne({var_name: "DefNodeURI", uri: newURI});
    }
    else
      await this.varsC.updateOne({var_name: "DefNodeURI" }, { $set: { uri: newURI }});
    return true;
  }

  async getAllNodeURIs() {
    let nodes = await (await this.nodesC.find({})).toArray();
    
    let outURIs = [];
    for (var i = 0; i < nodes.length; i++)
      outURIs[i] = nodes[i]._id.id;
    
    return outURIs;
  }

  async getAllNodesData() {
    let nodes = await (await this.nodesC.find({})).toArray();
    
    return nodes;
  }

  // Global data base functions
  async clearDB() {
    await this.nodesC.deleteMany({});
    await this.connectionsC.deleteMany({});
    await this.varsC.deleteMany({});

    return true;
  }

  async getDB() {
    return {
      nodes: await this.nodesC.find({}).toArray(),
      connections: await this.connectionsC.find({}).toArray(),
      variables: await this.varsC.find({}).toArray(),
    };
  }

  async loadDB( newDB ) {
    // Clear Collection
    await this.clearDB();
    // Load nodes
    if (newDB.nodes.length > 0)
      await this.nodesC.insertMany(recreateID(newDB.nodes));
    // Load connections
    if (newDB.connections.length > 0)
      await this.connectionsC.insertMany(recreateID(newDB.connections));
    // Load variables
    if (newDB.variables.length > 0)
      await this.varsC.insertMany(recreateID(newDB.variables));
    
    return true;
  }

  async addDB( newDB ) {
    // Load nodes
    if (newDB.nodes.length > 0)
      await this.nodesC.insertMany(recreateID(newDB.nodes));
    // Load connections
    if (newDB.connections.length > 0)
      await this.connectionsC.insertMany(recreateID(newDB.connections));
    
    return true;
  }
}

export class MongoDB { // Nodes data base
  dbs;
  mongodbConnection;

  async init( mongodbURL: string, dbNames: string[] ) {
    this.mongodbConnection = await MongoClient.connect(mongodbURL);
    var dbs = [];

    await Promise.all(dbNames.map(async ( name: string )=>{
     const db = new Database();
 
     await db.init(await this.mongodbConnection.db(name));
     dbs.push({ [name]: db });
    }));

    this.dbs = {};

    dbs.map((e)=>{
      this.dbs = {...this.dbs, ...e};
    });
  }

  async addDB( name: string ) {
    if (this.dbs[name] == undefined)
      return;
    const db = new Database();
    await db.init(await this.mongodbConnection.db(name));

    this.dbs = {
      ...this.dbs,
      [name]: db
    };
  }
}

File: server-rs/src/database/mongodb.rs
use super::{Node, NodeSkysphereData, Uri};

use base64::prelude::*;

pub struct MongoDB {
    nodes: mongodb::Collection<mongodb::bson::Document>,
    connections: mongodb::Collection<mongodb::bson::Document>,
    variables: mongodb::Collection<mongodb::bson::Document>,
}

impl MongoDB {
    pub async fn new(uri: &str) -> Option<Self> {
        let client =
            mongodb::Client::with_options(mongodb::options::ClientOptions::parse(uri).await.ok()?)
                .ok()?;
        let database = client.database("pml30map");

        // Some()
        Some(Self {
            nodes: database.collection::<mongodb::bson::Document>("nodes"),
            connections: database.collection::<mongodb::bson::Document>("connections"),
            variables: database.collection::<mongodb::bson::Document>("variables"),
        })
    }

    /// All node connections getting function
    async fn get_node_connections_filter(
        &self,
        filter: Option<mongodb::bson::Document>,
    ) -> Option<Vec<(Uri, Uri)>> {
        let mut cursor = self.connections.find(filter, None).await.ok()?;
        let mut data = Vec::<(Uri, Uri)>::new();

        loop {
            if let Some(v) = raw_document_to_connection(cursor.current()) {
                data.push(v)
            }

            if !cursor.advance().await.unwrap_or(false) {
                break;
            }
        }

        Some(data)
    }
}

impl Into<mongodb::bson::Document> for Node {
    fn into(self) -> mongodb::bson::Document {
        mongodb::bson::doc! {
            "_id": mongodb::bson::oid::ObjectId::from(self.uri.bytes()),
            "name": self.name,
            "position": {
                "x": self.position.0,
                "y": self.position.1,
                "z": self.position.2,
            },
            "skysphere": {
                "path": self.skysphere.path,
                "rotation": self.skysphere.rotation,
            },
            "floor": self.floor,
        }
    }
}

impl From<&mongodb::bson::RawDocumentBuf> for Node {
    fn from(value: &mongodb::bson::RawDocumentBuf) -> Self {
        Self {
            floor: value.get_i32("floor").unwrap_or(0),
            name: value.get_str("name").unwrap_or("").to_owned(),
            position: value
                .get_document("position")
                .map_or((0.0, 0.0, 0.0), |position| {
                    (
                        position.get_f64("x").unwrap_or(0.0) as f32,
                        position.get_f64("y").unwrap_or(0.0) as f32,
                        position.get_f64("z").unwrap_or(0.0) as f32,
                    )
                }),
            skysphere: value.get_document("skysphere").map_or(
                NodeSkysphereData {
                    path: "".into(),
                    rotation: 0.0,
                },
                |data| NodeSkysphereData {
                    path: data.get_str("path").unwrap_or("").to_owned(),
                    rotation: data.get_f64("rotation").unwrap_or(0.0) as f32,
                },
            ),
            uri: value
                .get_object_id("_id")
                .map(|v| Uri::from_bytes(v.bytes()))
                .unwrap_or(Uri::default()),
        }
    }
}

impl Into<mongodb::bson::oid::ObjectId> for Uri {
    fn into(self) -> mongodb::bson::oid::ObjectId {
        mongodb::bson::oid::ObjectId::from(self.bytes())
    }
}

impl From<mongodb::bson::oid::ObjectId> for Uri {
    fn from(value: mongodb::bson::oid::ObjectId) -> Self {
        Self::from_bytes(value.bytes())
    }
}

fn raw_document_to_connection(doc: &mongodb::bson::RawDocument) -> Option<(Uri, Uri)> {
    Some((
        doc.get_str("id1").ok().map(Uri::from_str).flatten()?,
        doc.get_str("id1").ok().map(Uri::from_str).flatten()?,
    ))
}

#[async_trait::async_trait]
impl super::Database for MongoDB {
    /// Node by uri getting function
    async fn get_node(&self, uri: Uri) -> Option<Node> {
        self.nodes
            .find_one(
                Some(
                    mongodb::bson::doc! { "_id": Into::<mongodb::bson::oid::ObjectId>::into(uri), },
                ),
                None,
            )
            .await
            .ok()
            .flatten()
            .map(|v| {
                let raw = mongodb::bson::RawDocumentBuf::from_document(&v)
                    .ok()?
                    .to_raw_document_buf();
                Some(Into::<Node>::into(&raw))
            })
            .flatten()
    }

    /// Node in database updating function
    async fn update_node(&mut self, uri: Uri, new_data: Node) -> Option<()> {
        self.nodes
            .update_one(
                mongodb::bson::doc! { "_id": Into::<mongodb::bson::oid::ObjectId>::into(uri), },
                Into::<mongodb::bson::document::Document>::into(new_data),
                None,
            )
            .await
            .map(|_| ())
            .ok()
    }

    /// Node delete function
    async fn delete_node(&mut self, uri: Uri) -> Option<()> {
        self.nodes
            .delete_one(
                mongodb::bson::doc! { "_id": Into::<mongodb::bson::oid::ObjectId>::into(uri), },
                None,
            )
            .await
            .map(|_| ())
            .ok()
    }

    /// Node adding function
    async fn add_node(&mut self, data: Node) -> Option<Uri> {
        self.nodes
            .insert_one(Into::<mongodb::bson::Document>::into(data), None)
            .await
            .ok()
            .map(|v| v.inserted_id.as_object_id())
            .flatten()
            .map(|v| Uri::from(v))
    }

    /// All node connections getting function
    async fn get_node_connections(&self, uri: Uri) -> Option<Vec<(Uri, Uri)>> {
        let a1 = self
            .get_node_connections_filter(Some(
                mongodb::bson::doc! {"id1" :Into::<mongodb::bson::oid::ObjectId>::into(uri)},
            ))
            .await;
        let a2 = self
            .get_node_connections_filter(Some(
                mongodb::bson::doc! {"id2" :Into::<mongodb::bson::oid::ObjectId>::into(uri)},
            ))
            .await;

        a1.zip(a2).map(|mut v| {
            v.0.append(&mut v.1);
            v.0
        })
    }

    /// Full connection list getting function
    async fn get_all_connections(&self) -> Option<Vec<(Uri, Uri)>> {
        self.get_node_connections_filter(None).await
    }

    // Node neighbour set getting function
    async fn get_node_neighbours(&self, uri: Uri) -> Option<Vec<Uri>> {
        self.get_node_connections(uri).await.map(|arr| {
            arr.into_iter()
                .map(|v| if uri != v.0 { v.0 } else { v.1 })
                .collect::<Vec<Uri>>()
        })
    }

    // Connection set getting function
    async fn add_connection(&mut self, uri1: Uri, uri2: Uri) -> Option<()> {
        self.connections
            .insert_one(
                mongodb::bson::doc! {
                    "id1": Into::<mongodb::bson::oid::ObjectId>::into(uri1),
                    "id2": Into::<mongodb::bson::oid::ObjectId>::into(uri2),
                },
                None,
            )
            .await
            .ok()
            .map(|_| ())
    }

    // Default node getting function
    async fn get_default_node_uri(&self) -> Option<Uri> {
        self.variables
            .find_one(Some(mongodb::bson::doc! { "var_name": "DefNodeURI" }), None)
            .await
            .ok()
            .flatten()
            .map(|doc| {
                doc.get_binary_generic("uri")
                    .ok()
                    .map(|v| Uri::from_slice(v.as_slice()))
            })
            .flatten()
            .flatten()
    }

    async fn set_default_node_uri(&mut self, new_uri: Uri) -> Option<()> {
        self.variables
            .update_one(
                mongodb::bson::doc! { "var_name": "DefNodeURI" },
                mongodb::bson::doc! {
                    "uri": mongodb::bson::binary::Binary::from_base64(BASE64_STANDARD.encode(new_uri.bytes().as_slice()), None).ok()?
                },
                None,
            )
            .await
            .ok()
            .map(|_| ())
    }

    async fn get_all_nodes(&self) -> Option<Vec<Node>> {
        let mut cursor = self.nodes.find(None, None).await.ok()?;
        let mut nodes: Vec<Node> = Vec::new();

        loop {
            nodes.push(Into::<Node>::into(&cursor.current().to_raw_document_buf()));

            if !cursor.advance().await.unwrap_or(false) {
                break;
            }
        }

        Some(nodes)
    }

    async fn get_all_node_uris(&self) -> Option<Vec<Uri>> {
        let mut cursor = self.nodes.find(None, None).await.ok()?;
        let mut uris: Vec<Uri> = Vec::new();

        loop {
            if let Ok(objid) = cursor.current().get_object_id("_id") {
                uris.push(Uri::from(objid));
            }

            if !cursor.advance().await.unwrap_or(false) {
                break;
            }
        }

        Some(uris)
    }
}
